cdap-security/src/main/java/co/cask/cdap/security/securestore/FileSecureStoreProvider.java

```diff
@@ -0,0 +1,543 @@
+        keyStore.deleteEntry(name);
+        throw k;
+      }
+      cache.put(metaKey, meta);
+    } catch (KeyStoreException e) {
+      throw new IOException("Failed to store the key. ", e);
+    } finally {
+      writeLock.unlock();
+    }
+    // Attempt to persist the store.
+    flush();
+  }
+
+  /**
+   * Deletes the element with the given name.
+   * @param name
+   */
+  void delete(String name) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    writeLock.lock();
+    if (cache.containsKey(metaKey)) {
+      cache.remove(metaKey);
+    }
+    try {
+      if (keyStore.containsAlias(name)) {
+        keyStore.deleteEntry(name);
+      }
+      /* It's OK if we were able to delete the data but not the metadata.
+         Java keystore allows overwrites, so the next time we try to write a
+         key with the same name the metadata will be overwritten. */
+      if (keyStore.containsAlias(metaKey)) {
+        keyStore.deleteEntry(metaKey);
+      }
+    } catch (KeyStoreException e) {
+      LOG.error("Failed to delete the key " + name, e);
+    } finally {
+      writeLock.unlock();
+    }
+    // Attempt to persist the store.
+    flush();
+  }
+
+  /**
+   * List of all the entries in the secure store.
+   * @return A map of name -> description
+   * @throws IOException
+   */
+  Map<String, String> list() throws IOException {
+    Map<String, String> list = new HashMap<>();
+    String name = null;
+    readLock.lock();
+    try {
+      try {
+        Enumeration<String> aliases = keyStore.aliases();
+        // Surprisingly enhanced for-each does not work for Enumeration
+        while (aliases.hasMoreElements()) {
+          name = aliases.nextElement();
+          // We don't want to list the meta data elements separately.
+          if (name.endsWith(METADATA_SUFFIX)) {
+            continue;
+          }
+          list.put(name, getDescription(name));
+        }
+      } catch (KeyStoreException e) {
+        throw new IOException("Failed to get the description for " + name, e);
+      }
+      return list;
+
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  /**
+   * Returns the metadata for the element identified by the given name.
+   * @param name
+   * @return
+   * @throws IOException
+   */
+  SecureStoreMetadata getSecureStoreMetadata(String name) throws IOException {
+    String metaKey = constructMetadataKey(name);
+    if (cache.containsKey(metaKey)) {
+      return cache.get(metaKey);
+    }
+    try {
+      if (!keyStore.containsAlias(metaKey)) {
+        throw new IOException("Metadata for " + name + " not found in the secure store.");
+      }
+      SecureStoreMetadata meta = ((KeyMetadata) keyStore.getKey(metaKey, password)).metadata;
+      cache.put(metaKey, meta);
+      return meta;
+    } catch (NoSuchAlgorithmException | UnrecoverableKeyException | KeyStoreException e) {
+      throw new IOException("Unable to retrieve the metadata for " + name);
```

Comment:
Should still pass the `e` as the exception cause, otherwise it would be difficult to debug. Also, the client here is actually CDAP, so there is no point in hiding.